### ISSUE TITLE
Spike: Define upgrade path to use sources-api v3 HTTP endpoints

### ISSUE THREAD

**Issue by katherine-black** [cloudigrade/cloudigrade#716]
In GitLab by @infinitewarp on Jul 15, 2020, 09:26

## Summary

As of the time of this writing, sources-api offers a v3 HTTP API, but cloudigrade still uses a mix of v1 and v2 endpoints. We need to understand what behaviors will change if we start using all v3 APIs and if what we need to change to handle those new behaviors.

Output:
- [ ] Shared notes/documentation that explains the difference in behavior for us when we change each of our v1 or v2 use cases to v3 HTTP.
- If changes appear to be relatively simple:
  - [ ] submit a MR for cloudigrade including these changes.
  - [ ] QE to rerun tests to check for regressions
- else:
  - [ ] define a new user story issue for cloudigrade, and link it here.

Time box: 2 days

## Assumptions and Questions

- Be sure to check the flows around:
  - Creating sources-api objects resulting in cloudigrade using sources-api to create/update the cloud account
  - When cloudigrade disables a CloudAccount due to some error, cloudigrade uses sources-api to push an "unavailable" status.
  - If there are other noteworthy flows, please document them here as you find them.

**Comment by katherine-black** [comment]
In GitLab by @infinitewarp on Jul 28, 2020, 09:56

changed title from **Spike: Define upgrade path to use sources-api v3 endpoints** to **Spike: Define upgrade path to use sources-api v3 {+HTTP +}endpoints**

**Comment by katherine-black** [comment]
In GitLab by @werwty on Jul 28, 2020, 16:36

Jupyter notebook from poking sources [710-sources-v3.ipynb](/uploads/d5b82bca3032cc3b4240c23d3bb3c0a2/710-sources-v3.ipynb)

TLDR:
A diff of the v2.0 and v3.0 openapi spec shows that there is very little difference between the two specs. 

The one change that would affect us is that Authentication `authtype` is now marked as `readonly`. We currently create a cloud meter account only if this field is set to `cloud-meter-arn`. Despite being marked `readonly` in the spec the API call with this field set still works. But QE tests might break since they depend on the client generated from the spec file.

It also looks like they mostly fixed the bug where cascade delete did not fire messages. I say mostly because it looks like the ApplicationAuthentication.destroy message still isn't being created on a cascade delete. Which is a pity, because that's the one that would allow us to simplify our sources listener workflow. 
See: https://github.com/RedHatInsights/sources-api/issues/179#issuecomment-665242567

One last thing I noticed is that create and delete messages oftentimes arrive out of order. Our account creation code handles this just fine; if a source (or any of its constituents) is deleted, we detect it during account creation and do not create the clount. But this does affect running tests, since if there isn't enough time between sources creation and deletion, we do not end up creating that account at all.

**Comment by katherine-black** [comment]
In GitLab by @pakamble on Jul 30, 2020, 04:55

Started regression runs to validate this story. I will update again here  once all tests runs get completed.

**Comment by katherine-black** [comment]
In GitLab by @pakamble on Jul 31, 2020, 03:49

Regression test suite passed cleanly no issue found in sources API.
[sources_regression_test_logs.log](/uploads/5ffdb07310bfe66101e2fff49cc785cb/sources_regression_test_logs.log)
